### ISSUE TITLE
Build error NETSDK1124 with NET8 RC2 when multiple TargetFrameworks

### ISSUE THREAD

**Issue by snakefoot** [dotnet/runtime#94406]
### Description

Trying to publish an application `<TargetFramework>net8.0</TargetFramework>` with trimming enabled:
```
dotnet publish -r win-x64 -c release --self-contained -p:PublishTrimmed=true .\examples\NetCore2\HostingExample
```
The application depends on two libraries:
- **NLog.Extensions.Logging**:
  ```
  <TargetFrameworks>net461;netstandard1.3;netstandard1.5;netstandard2.0;netstandard2.1;net6.0;net8.0</TargetFrameworks>
  ```
- **NLog.Extensions.Hosting**:
  ```
  <TargetFrameworks>netstandard2.0;netstandard2.1;net6.0;net8.0</TargetFrameworks>
  ```

The build fails with the following error:
```
C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceRes
olution.targets(90,5): error NETSDK1124: Trimming assemblies requires .NET Core 3.0 or higher. [C:\Users\snakefoot\sour
ce\repos\NLog.Extensions.Logging\src\NLog.Extensions.Hosting\NLog.Extensions.Hosting.csproj::TargetFramework=netstandar
d2.0]
C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceRes
olution.targets(90,5): error NETSDK1124: Trimming assemblies requires .NET Core 3.0 or higher. [C:\Users\snakefoot\sour
ce\repos\NLog.Extensions.Logging\src\NLog.Extensions.Logging\NLog.Extensions.Logging.csproj::TargetFramework=net461]
```

### Reproduction Steps

Publish NET8 application with trimming, that depends on projects with mutiple `<TargetFrameworks>` (Besides NET8)

Also tried to update library-projects to be very explicit about when to enable trimming, but build with publish still fails (No warnings or errors with standard `dotnet build`):
```
    <EnableTrimAnalyzer Condition=" '$(TargetFramework)' == 'net8.0'">true</EnableTrimAnalyzer>
    <IsTrimmable Condition=" '$(TargetFramework)' == 'net8.0'">true</IsTrimmable>
```

Also tried adding these command-line options `-f net8.0` or `-p:TargetFramework=net8.0`, but the build-error remains.

### Expected behavior

Build publishes the application without errors.

### Actual behavior

Build fails with `error NETSDK1124: Trimming assemblies requires .NET Core 3.0 or higher.` for the library projects.

### Regression?

Build publish with trimming works fine with NET6 + NET7

### Known Workarounds

Uninstall NET8 RC2 SDK or stop using multiple TargetFrameworks in library-projects or depend on nuget-packages instead of the library-projects directly.

**Comment by ViktorHofer** [comment]
Good point. I just moved the issue back. cc @agocke

**Comment by ChaosEngine** [comment]
I am seeing similiar error on
`dotnet publish --no-restore -c Release --self-contained -r linux-x64  -p:PublishTrimmed=true Project`
when including `<TargetFramework>netstandard2.0</TargetFramework>` library as dependency of main ` <TargetFramework>net8.0</TargetFramework` build

**Comment by ViktorHofer** [comment]
cc @vitek-karas

**Comment by vitek-karas** [comment]
@snakefoot, could you please try adding `<PublishTrimmed>true</PublishTrimmed>` to your project file for the app and remove it from the command line?

There is a somewhat unexpected MSBuild behavior where all properties passed on the command line are forced onto all projects (even dependencies), and `PublishTrimmed` will turn on `EnableTrimAnalyzers` which will cause the error on projects targeting just `netstandard2.0` (for a good reason). This behavior makes littles sense for property like `PublishTrimmed`, but MSBuild currently does it for everything.

/cc @sbomer

**Comment by snakefoot** [comment]
Yes moving `<PublishTrimmed>true</PublishTrimmed>` into the application csproj-file resolved the publish error for me.

Combined with adding `<IsTrimmable Condition="'$(TargetFramework)' == 'net8.0'">true</IsTrimmable>` to the other library-projects.

**Comment by ChaosEngine** [comment]
> @snakefoot, could you please try adding `<PublishTrimmed>true</PublishTrimmed>` to your project file for the app and remove it from the command line?
> 
> There is a somewhat unexpected MSBuild behavior where all properties passed on the command line are forced onto all projects (even dependencies), and `PublishTrimmed` will turn on `EnableTrimAnalyzers` which will cause the error on projects targeting just `netstandard2.0` (for a good reason). This behavior makes littles sense for property like `PublishTrimmed`, but MSBuild currently does it for everything.
> 
> /cc @sbomer

My problem is also resolved for two Docker builds (alpine and debian based) with trimming enabled - now only in top-most *.csproj I am having `<PublishTrimmed>true</PublishTrimmed>

**Comment by vitek-karas** [comment]
Our guidance is to put the `PublishTrimmed` (or `PublishAot`) properties into the project file. Not only it doesn't have this particular problem but it will also:
* enable analyzers for your app's project so that you get trim/AOT related warnings during build
* modifies some settings such that the app will behave the same in debug builds as well as when published trimmed (trimming by default disables some runtime functionality which is not compatible with trimming for example)
* this means that you should gain confidence in the behavior of the app after publish, and you can test with the same runtime behavior it will have after trimming.

@sbomer - can we raise this again with the SDK/MSBuild people, to see if we can make progress on it? Originally it was just a quirk we found internally, but this indicates people are running into it.

**Comment by snakefoot** [comment]
I think the current [build-error](https://learn.microsoft.com/en-us/dotnet/core/tools/sdk-errors/) message is a little confusing:
```
NETSDK1124 | Trimming assemblies requires .NET Core 3.0 or higher.
```
When the solution involves removing `PublishTrimmed` from command-line and instead specifying it in app-csproj-file, and also making sure `IsTrimmable` is conditional.

**Comment by vitek-karas** [comment]
The problem is that the error was never meant to be triggered by `PublishTrimmed`, it's there if you add `IsTrimmable` onto a `netstandard2.0` project. But the flow of the various properties causes it to show up anyway. I guess we could detect that `PublishTrimmed` is set and show a different error possibly.

**Comment by itn3000** [comment]
I found small reproducible steps;

1. `dotnet new console --name app1`
2. `dotnet new classlib --name lib1 -f netstandard2.0`
3. `dotnet add app1\app1.csproj reference lib1\lib1.csproj`
4. `dotnet publish app1\app1.csproj -r win-x64 --self-contained -p:PublishTrimmed=true`

then get NETSDK1124 error.

I also tested on dotnet-sdk 6.0.417 and 7.0.203(using global.json to change SDK), I got no error.

**Comment by agocke** [comment]
Another way to easily handle this in the project file is to use a condition:

```xml
  <PropertyGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))">
    <PublishTrimmed>true</PublishTrimmed>
  </PropertyGroup>
```
